### PR TITLE
roachtest: optimize cluster destroy and create

### LINKED ISSUE #125228: roachtest: optimize cluster destroy and create
Currently, roachtest creates a cluster for running tests. Once the test is complete, it tries to reuse the cluster for the next test. Incase case the cluster cannot be reused for the next test, it destroys and creates a new cluster. This happens in sequence today.
cluster destory takes around a minute or two and the same is with cluster create. So, overall, this operation takes away ~4 minutes of the roachtest time.
If we run destroy in a goroutine and start the cluster create for the next test, we should be able to save 2 minutes.

Considering a teamcity nightly run, we see that we create/destroy ~20 clusters in a worker, which accounts to 2*20 = 40 minutes!!

Jira issue: CRDB-39323

### PR DESCRIPTION
**Pull request by nameisbhaskar**
Currently, roachtest creates a cluster for running tests. Once the test is complete, it tries to reuse the cluster for the next test. Incase case the cluster cannot be reused for the next test, it destroys and creates a new cluster. This happens in sequence today.
cluster destory takes around a minute or two and the same is with cluster create. So, overall, this operation takes away ~4 minutes of the roachtest time.
If we run destroy in a goroutine and start the cluster create for the next test, we should be able to save 2 minutes.

Considering a teamcity nightly run, we see that we create/destroy ~20 clusters in a worker, which accounts to 2*20 = 40 minutes!!

Fixes: #125228
Epic: None

### CODE REVIEW

File: pkg/cmd/roachtest/test_runner.go
Comment (herkolategan) [inline pkg/cmd/roachtest/test_runner.go:625]: It seems like the `waitgroup wait` is added before all possible return points. Is there any reason for not moving it a `defer` above, and creating the waitgroup a little earlier so it can be part of that `defer`, so if someone adds another return point they wouldn't have to keep this in mind? Similar to how the `quota` is release.

File: pkg/cmd/roachtest/test_runner.go
Comment (DarrylWong) [inline pkg/cmd/roachtest/test_runner.go:946]: Is there a case where the cluster would be nil?

File: pkg/cmd/roachtest/test_runner.go
Comment (DarrylWong) [inline pkg/cmd/roachtest/test_runner.go:618]: nit: if you do keep the `wait` flag, then add a `/* Wait */` comment.

File: pkg/cmd/roachtest/test_runner.go
Comment (DarrylWong) [inline pkg/cmd/roachtest/test_runner.go:951]: I don't quite follow this comment. Is it because you can only have one local cluster so it must be done sequentially?

File: pkg/cmd/roachtest/test_runner.go
Comment (DarrylWong) [inline pkg/cmd/roachtest/test_runner.go:944]: Instead of passing a bool to cover this edge case, what if you just called `c.Destroy(context.Background(), closeLogger, l)` directly when you don't want to wait?

File: pkg/cmd/roachtest/test_runner.go
Comment (DarrylWong) [inline pkg/cmd/roachtest/test_runner.go:608]: nit: exit the roachtest runner?

**Review by DarrylWong — COMMENTED** [review]
Nice! I do wonder if we will run into any issues wrt quota if we're now trying to create resources before they are freed on the cloud end. Maybe the built in retry would be enough time to get past that though.

Should we also run a smoke test before merging?

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go:946]: This is a prcaution, in case the cluster is passed as nil. Currently, the code path should hit this condition.

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go:951]: Yes, without this condition CI fails as it uses local cluster.

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go]: wait is removed as per your suggestion

**Comment by nameisbhaskar** [comment]
> Nice! I do wonder if we will run into any issues wrt quota if we're now trying to create resources before they are freed on the cloud end. Maybe the built in retry would be enough time to get past that though.
> 
> Should we also run a smoke test before merging?

I ran a nightly build from this. I will run another one and paste the same here!

**Comment by nameisbhaskar** [comment]
Thanks @herkolategan and @DarrylWong for reviewing this! I have addressed the comments. Please do take a look.

I will run one nightly build from this PR to see all runs ok.

**Comment by nameisbhaskar** [comment]
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15600985?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go:942]: There was one flow in which there was a possibility of cluster being nil because of a if condition, where I added this. The scenario is no more now, but, I kept it just in case.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:618]: What's the rationale for adding a blocking call on these paths? Note that generally speaking this call has no upper bound on the running time.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:608]: Nit: comment can be made more precise. The `WaitGroup` just ensures that all pending Destroy calls have completed.

However, note that this may be introducing a new behavior; whereas previously an interrupted worker would exit abruptly, now it has to wait for Destroy(s) to finish.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [reply pkg/cmd/roachtest/test_runner.go:942]: If it's _provable_ that `c` cannot be nil, then it's best to make it a _precondition_.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:956]: Nit: this seems redundant. We already log completion in `cluster.go` (`doDestroy`).

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go]: This is correct. Till today, we were relying on GC to destroy the clusters in case of errors.

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go]: Sure!

File: pkg/cmd/roachtest/test_runner.go
Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go]: makes sense. Will remove this and rely of GC to handle this.

**Comment by nameisbhaskar** [comment]
Thanks @srosenberg for the review.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:723]: I'm still fuzzy why we're adding `c.Destroy` to error paths? Don't we want to fail fast instead and let GC handle the clean-up at a later time?

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:608]: Nit: a single line comment should suffice. What part of this code isn't self-documenting? Arguably, only the `defer` statement may need a supporting comment.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [reply pkg/cmd/roachtest/test_runner.go:936]: Let's state that explicitly in the comment; instead of the above, something like this,

> N.B. multiple local clusters aren't supported, hence we must use a blocking call.

File: pkg/cmd/roachtest/test_runner.go
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:935]: I'd factor this case out and rename the above func to something like `destroyClusterAsync`. Otherwise, there is a confounding factor. The above comment (L666) says, "do not wait", yet we end up waiting in case of `c.IsLocal()`. Also, the `WaitGroup` is coupled with the case which doesn't need it, since it's a blocking call.

Comment (nameisbhaskar) [reply pkg/cmd/roachtest/test_runner.go:935]: makes sense! I updated the comments. Thanks @srosenberg !

**Review by DarrylWong — APPROVED** [review]
LGTM

**Review by srosenberg — APPROVED** [review]
Let's do another, final smoke test run with SELECT_PROBABILITY=0.6, then merge.

**Comment by nameisbhaskar** [comment]
Thanks @DarrylWong and @srosenberg !

I have started the nightly with SELCT_PROBABILITY=0.6.
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/15637536?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

I will merge once this completes and all looks good!

**Comment by nameisbhaskar** [comment]
bors r=@DarrylWong,@srosenberg

**Pull request merged**
